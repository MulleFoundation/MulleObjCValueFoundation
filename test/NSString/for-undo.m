#ifndef __MULLE_OBJC__
# import <Foundation/Foundation.h>
#else
# import <MulleObjCValueFoundation/MulleObjCValueFoundation.h>
#endif

static void    print_c_d_flag( unichar c, unichar d, BOOL flag)
{
   mulle_printf( "\n\t\t");
   mulle_printf( (c <= 0) ? "%d": "'%lC'", c);
   mulle_printf( " ");
   mulle_printf( (d <= 0) ? "%d": "'%lC'", d);
   mulle_printf( " (%bd)", flag);
}


static void    test( NSString *s, NSUInteger n, NSUInteger m)
{
   struct MulleStringEnumerator   rover;
   NSUInteger   i, j;
   BOOL         flag;
   unichar      c, d;

   mulle_printf( "\"%#s\" (%td, %td)", [s UTF8String], n, m);

   _MulleStringEnumeratorInit( &rover, s);

   if( n)
   {
      mulle_printf( "\n\tnext:");

      for( i = 0; i < n; i++)
      {
         if( MulleStringEnumeratorNext( &rover, &c))
            mulle_printf( "\n\t\t'%lC' ", c);
      }
   }

   if( m)
   {
      mulle_printf( "\n\tundo:");

      for( j = 0; j < m; j++)
      {
         flag = _MulleStringEnumeratorUndoNext( &rover, &c, &d);
         print_c_d_flag( c, d, flag);
      }
   }

   mulle_printf( "\n\tnext:");
      if( MulleStringEnumeratorNext( &rover, &c))
         mulle_printf( "\n\t\t'%lC' ", c);

   MulleStringEnumeratorDone( &rover);

   mulle_printf( "\n\n");
}



int   main( void)
{
   NSUInteger   i, j;

   for( i = 1; i < 4; i++)
      for( j = 1; j < 4; j++)
         test( @"VfL", i, j);

   return( 0);
}
