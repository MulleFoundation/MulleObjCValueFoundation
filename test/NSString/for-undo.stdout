"VfL" (1, 1)
	next:
		'V' 
	undo:
		'V' 0 (YES)
	next:
		'V' 

"VfL" (1, 2)
	next:
		'V' 
	undo:
		'V' 0 (YES)
		-1 0 (NO)
	next:
		'V' 

"VfL" (1, 3)
	next:
		'V' 
	undo:
		'V' 0 (YES)
		-1 0 (NO)
		-1 0 (NO)
	next:
		'V' 

"VfL" (2, 1)
	next:
		'V' 
		'f' 
	undo:
		'f' 'V' (YES)
	next:
		'f' 

"VfL" (2, 2)
	next:
		'V' 
		'f' 
	undo:
		'f' 'V' (YES)
		'V' 0 (YES)
	next:
		'V' 

"VfL" (2, 3)
	next:
		'V' 
		'f' 
	undo:
		'f' 'V' (YES)
		'V' 0 (YES)
		-1 0 (NO)
	next:
		'V' 

"VfL" (3, 1)
	next:
		'V' 
		'f' 
		'L' 
	undo:
		'L' 'f' (YES)
	next:
		'L' 

"VfL" (3, 2)
	next:
		'V' 
		'f' 
		'L' 
	undo:
		'L' 'f' (YES)
		'f' 'V' (YES)
	next:
		'f' 

"VfL" (3, 3)
	next:
		'V' 
		'f' 
		'L' 
	undo:
		'L' 'f' (YES)
		'f' 'V' (YES)
		'V' 0 (YES)
	next:
		'V' 

